src/team_lib/mhba/xccl_mhba_mkeys.c

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#include "xccl_mhba_lib.h"
+
+
+//todo extract to helper header
+int poll_cq(struct ibv_cq *cq, int expected_wc_num, int *actual_wc_num, struct ibv_wc actual_wc[]) {
+    *actual_wc_num = -1; // Initial value.
+    long int poll_for_completion_retry = 10;
+    int total_received_completions = 0;
+    int new_received_completions = 0;
+    int completions_to_wait_for = expected_wc_num;
+
+    // TODO: Consider doing this instead of requiring the user to initialize
+    // the array of completions.
+    //memset((void*)actual_wc, 0, expected_wc_num*sizeof(struct ibv_wc));
+
+    while ((completions_to_wait_for > 0) && (poll_for_completion_retry > 0)) {
+        new_received_completions = ibv_poll_cq(cq, completions_to_wait_for, &actual_wc[total_received_completions]);
+        if (new_received_completions < 0) {
+            xccl_mhba_error("ibv_poll_cq() failed");
+            return new_received_completions;
+        }
+        total_received_completions += new_received_completions;
+        completions_to_wait_for -= new_received_completions;
+        if (completions_to_wait_for > 0) {
+            poll_for_completion_retry--;
+            sleep(1); // todo ?
```

Comment:
no sleeps in the code

Reply:
from Rami's code, this is why i left the todo - need to check with him. i'll leave it for now, will be addressed on next PR